feat(formatting): Remove unused parent tables

iepathos opened the pull request:
Implements TOML formatting enhancement for #32

## Summary

Removes empty parent table headers that:
- Have no key-value content
- Have no comments in their range
- Have a child table defined

Before:                                                    
```toml
[a]
[a.b]
[c]

[d] # Hello
[d.e]
```

After:
```toml                                                  
[a.b]
[c]

[d] # Hello
[d.e]
```
Implementation

Two-pass algorithm in src/formatting/unused_parents.rs:

1. Collect tables: Find all table headers, walk backwards to include adjacent leading comments in start, set end to next table's header index (or EOF)
2. Remove unused parents: For each table that is a prefix of another, check if its range contains any non-whitespace content outside the header

Uses a Table struct with name, start, end, is_array_table fields. Array tables ([[table]]) are always preserved because removing them would change semantics.
                                                           
Test coverage
33 unit tests covering:
- Single/multiple/deeply nested empty parents
- Leading comments (adjacent vs separated by blank line)
- Trailing comments on header line
- Parents with key-value content (preserved)
- Standalone tables without children (preserved)
- Array tables as parents and children (preserved)
- Quoted table names
- Whitespace handling between tables

epage commented:
> ## Implementation
> 
> Two-pass algorithm in `src/formatting/unused_parents.rs`:
> 1. Collect metadata for all tables (name segments, token range, has content, has comment)
> 2. Remove standard tables that are empty, uncommented prefixes of other tables

As a heads up, my assumption is that future PRs will generalize the metadata collection into a second representation for additional table manipulation operations like sorting.

iepathos commented:
Rebased main to address merge conflict.

Reorganized commits in this PR so we have a no-op formatter with tests demonstrating the current behavior in the first commit and the following commit has the implementation and test expectation updates.

Should we consider different organization for the metadata collection with future changes in mind or keep it in unused_parents.rs for now?

epage commented:
Some more test cases for us to consider:

```toml
[parent.child]

[parent]
```

```toml
# leading comment
[parent]
[parent.child]
```

```toml
# body comment

[parent]
[parent.child]
```

```toml
[other]

# leading comment
[parent]
[parent.child]
```

```toml
[other]
# ambiguous comment
[parent]
[parent.child]
```

```toml
[other]

# body comment

[parent]
[parent.child]
```

```toml
[parent]
# body comment

[parent.child]
```

```toml
[parent]

# body comment

[parent.child]
```


```toml
[parent]
# ambiguous comment
[parent.child]
```

epage commented:
Another test case

```toml
[parent]

[parent.child]
```
to show what happens to whitespace when it is removed.

epage commented on src/formatting/unused_parents.rs:
```diff
@@ -1,8 +1,168 @@
+    let mut name = Vec::new();
+    let mut i = start;
+
+    while i < tokens.len() {
```
How well would using `TokenIndices` work?

iepathos replied:
It would work well and be more consistent with other formatter implementations in the codebase. Will refactor to use TokenIndices.

epage commented on src/formatting/unused_parents.rs:
```diff
@@ -1,8 +1,168 @@
+            k if k == close_kind => {
+                return (name, i);
+            }
+            _ => break,
```
whats so different between these cases?

epage commented on src/formatting/unused_parents.rs:
```diff
@@ -434,7 +575,6 @@ y = 2
 ",
             str![[r#"
-[parent]
 # body comment
```
Isn't a comment like this part of `parent`, just like a comment on the same line?

epage commented on src/formatting/unused_parents.rs:
```diff
@@ -340,7 +486,6 @@ y = 2
             str![[r#"
 # leading comment
-[parent]
 [parent.child]
```
Isn't a comment like this part of `parent`, just like a comment on the same line?

epage commented on src/formatting/unused_parents.rs:
```diff
@@ -1,8 +1,168 @@
+    header_end: usize,
+    has_content: bool,
+    has_comment: bool,
+}
```
As a follow up to my heads up at https://github.com/crate-ci/cargo-cargofmt/pull/35#issuecomment-3693995993, my assumption for how this will look in the future is that it will be:
```rust
struct Table {
    name: Vec<String>,
    start: usize,
    end: usize,
}
```
where the start includes any leading comments and the end goes all the way to after any key-value pairs, comments, or trailing newlines.

In cases like this one, we could then walk the content to see if comments or values are present.  When sorting, we then know exactly what range to copy over into the new location.

As per before, do what works for now and we can evolve.  It might end up looking like a hybrid of the two or something completely different, depending on what we find out as we implement more formatting rules.  For instance, I didn't have a field for `is_array_table` but that probably should be there rather than scanning for it every time.

iepathos replied:
Yes, this will be better.  Will simplify with this approach

iepathos commented on src/formatting/unused_parents.rs:
```diff
@@ -378,7 +522,6 @@ y = 2
 
 # leading comment
-[parent]
 [parent.child]
```
Yes, to be safe we should consider a comment before the parent without blank lines between them as attached to the parent.

iepathos commented on src/formatting/unused_parents.rs:
```diff
@@ -1,8 +1,168 @@
+    let mut tables = Vec::new();
+    let mut i = 0;
+
+    while i < tokens.len() {
```
It works well. Updated to use TokenIndices

iepathos commented on src/formatting/unused_parents.rs:
```diff
@@ -471,7 +610,6 @@ y = 2
             str![[r#"
-[parent]
 # ambiguous comment
 [parent.child]
```
Updated implementation to consider any comment following a parent header before another header as attached to the parent regardless of blank lines to be safe with these ambiguous cases.

iepathos commented on src/formatting/unused_parents.rs:
```diff
@@ -396,7 +539,6 @@ y = 2
 [other]
 # ambiguous comment
-[parent]
 [parent.child]
```
We should consider a leading comment without blank lines between it and the parent header as attached to be safe. Updated implementation for this.

iepathos commented:
Updated the implementation to use these comment attachment rules for safer pruning:

1. A comment is on the same line as the parent header.
2. A comment immediately precedes a parent with no blank line separation.
3. Any comment following a parent and between it and other headers, i.e. body comments, should be considered attached to the parent regardless of blank lines.

These rules preserve parent tables in ambiguous situations.

Added prev_index for reversing through indices to TokenIndices. This is used by find_leading_blank_lines_start and is_preceded_by_adjacent_comment in unused_parents.rs. The TokenIndices change is separated out in two commits before the unused parent commits.

epage commented on src/toml/mod.rs:
> feat(toml): Add prev_index to TokenIndices 

Note that I put `feat` as applying to user-facing changes.  I'm fine with this function and tests being added in the commit that needs them.

epage commented on src/formatting/unused_parents.rs:
```diff
@@ -1,8 +1,303 @@
+    has_content: bool,
+    has_header_comment: bool,  // Comment on same line as header
+    has_leading_comment: bool, // Comment immediately before header (no blank line)
+    has_body_comment: bool,    // Any comment in body (after header, before next table)
```
Do we care about what kind of comment it is?

epage commented on src/formatting/unused_parents.rs:
```diff
@@ -1,8 +1,303 @@
+            for i in leading_start..=table.trailing_ws_end {
+                tokens.tokens[i] = TomlToken::EMPTY;
+            }
+        }
```
Looking at the test cases, this is changing the meaning of the comments
```toml
[other]

# body comment

[parent]
[parent.child]
```
should be
```toml
[other]

# body comment

[parent.child]
```
and not
```toml
[other]

# body comment
[parent.child]
```

iepathos commented on src/formatting/unused_parents.rs:
```diff
@@ -1,8 +1,303 @@
+    (name, tokens.len().saturating_sub(1))
+}
+
+fn find_header_end(tokens: &TomlTokens<'_>, close_idx: usize) -> (usize, bool) {
```
Function renamed for clarity

iepathos commented on src/formatting/unused_parents.rs:
```diff
@@ -1,8 +1,303 @@
+            TokenKind::Newline => {
+                last_ws = i;
+            }
+            _ => break,
```
Yes, added this

epage commented on src/formatting/unused_parents.rs:
```diff
@@ -0,0 +1,771 @@
+        .collect()
+}
+
+fn should_remove(
```
Please organize the file in call order e.g. this goes above the functions it calls

iepathos commented on src/formatting/unused_parents.rs:
```diff
@@ -0,0 +1,771 @@
+
+    for table in tables.iter().rev() {
+        if should_remove(table, tokens, &parent_names) {
+            for i in table.start..=table.end {
```
Will make the end exclusive to match default expectations and avoid needing extra comments for it

iepathos commented on src/formatting/unused_parents.rs:
```diff
@@ -0,0 +1,771 @@
+        return false;
+    }
+
+    if has_comment(tokens, table.start, table.end) {
```
Much simpler, thank you
